components/site/src/lib.rs

```diff
@@ -799,19 +799,30 @@ impl Site {
     }
 
     fn index_for_lang(&self, lang: &str) -> Result<()> {
-        let index_json = search::build_index(lang, &self.library.read().unwrap(), &self.config)?;
-        let (path, content) = match &self.config.search.index_format {
-            IndexFormat::ElasticlunrJson => {
-                let path = self.output_path.join(format!("search_index.{}.json", lang));
-                (path, index_json)
+        let extension = match &self.config.search.index_format {
+            IndexFormat::ElasticlunrJavascript | IndexFormat::FuseJavascript => "js",
+            IndexFormat::ElasticlunrJson | IndexFormat::FuseJson => "json",
+        };
+        let path = &self.output_path.join(format!("search_index.{}.{}", lang, extension));
+        let library = self.library.read().unwrap();
+        let content = match &self.config.search.index_format {
+            IndexFormat::ElasticlunrJavascript | IndexFormat::ElasticlunrJson => {
+                search::build_index(lang, &library, &self.config)?
             }
-            IndexFormat::ElasticlunrJavascript => {
-                let path = self.output_path.join(format!("search_index.{}.js", lang));
-                let content = format!("window.searchIndex = {};", index_json);
-                (path, content)
+            IndexFormat::FuseJson | IndexFormat::FuseJavascript => {
+                search::fuse::build_index(lang, &library, &self.config.search)?
             }
         };
-        create_file(&path, &content)
+        drop(library); // no need to hold on to this guard while writing
+        create_file(
+            path,
+            match self.config.search.index_format {
+                IndexFormat::ElasticlunrJson | IndexFormat::FuseJson => content,
+                IndexFormat::ElasticlunrJavascript | IndexFormat::FuseJavascript => {
+                    format!("window.searchIndex = {}", content)
```

Comment:
This can probably go on the enum as well?

Reply:
of course it could, but I feel like it would be a weird thing to do? completely subjective.

Reply:
like at that point i might as well put the entire thing inside IndexFormat